### ISSUE TITLE
[meteor 1.3-beta11] hot code push doesn't work in production build

### ISSUE THREAD

**Issue by markoshust** [meteor/meteor#6370]
Hot code push is working for me in TestFlight, but when app is published to the App Store, it doesn't work. Unfortunately, I don't have any more info for you as this only happens in the production build.

re: https://github.com/meteor/meteor/issues/6266

**Comment by rlech** [comment]
why is your testflight build not your testflight build?

**Comment by martijnwalraven** [comment]
Do both builds connect to the same server?

**Comment by markoshust** [comment]
Sorry, bad phrasing. It's the same build, same server.

It seems the 'release' tag has something to do with it, as testflight doesn't use the 'release' build.

**Comment by martijnwalraven** [comment]
Hmm, without more information I can only guess what could be going on.

Do you specify the server using an IP address or a domain name? (thinking it might have something to do with the ATS configuration)

**Comment by markoshust** [comment]
Sorry, I know I don't have much more to go on than you.

It's a domain name over SSL.

**Comment by martijnwalraven** [comment]
I'm sorry, but I don't think I can help you without more information.

**Comment by markoshust** [comment]
No problem, I understand. I'll try to get some more details later next week.... when I make another push to production. :laughing:

**Comment by markoshust** [comment]
@martijnwalraven ok, i've confirmed that hot code push doesn't seem to be working at all now. the last known confirmed working version was 1.3-beta8. i'm currently on 1.3-beta12.

deploys are running, public site is updating. however, the app is not.

**Comment by martijnwalraven** [comment]
How are you deploying and how has the server been configured? 
- Are you sure it is serving the Cordova code? Can you open `http://<server>/__cordova/`?
- Has the `ROOT_URL` been configured on the server?

Before, you mentioned TestFlight builds were working and connected to the same server. Is this still the case? Or is the difference only between local development and release builds?

**Comment by markoshust** [comment]
- I'm deploying with Docker & Kubernetes. Meteor sets behind an Nginx proxy for SSL termination.
- I'm not sure exactly what is supposed to happen when visiting http://<server>/__cordova (cordova plugins are working and etc.), however I'm receiving this error:
![js-error](https://cloud.githubusercontent.com/assets/437029/13553469/df6d0c88-e358-11e5-9a33-a36514134fd8.png)
- `ROOT_URL` is configured.
- TestFlight builds are no longer working. I believe this stopped working in beta11 (I believe I didn't notice this the first time they stopped working).

I'm assuming the js error has a lot to do with this, please let me know if something is misconfigured!

Cheers,
Mark

**Comment by markoshust** [comment]
@martijnwalraven in safari i'm getting this error, not sure if it's one in the same as the above

![screen shot 2016-03-06 at 5 08 23 am](https://cloud.githubusercontent.com/assets/437029/13553489/9ec1a2e2-e359-11e5-92b7-0fa913f42534.png)

**Comment by markoshust** [comment]
I've confirmed in my TestFlight build now that removing
`X-Content-Type-Options: nosniff`
from my Nginx config seems to have fixed this issue. I'll reopen if I have another issue. Thanks for your help!

Mark

**Comment by martijnwalraven** [comment]
Hmmm, that would be really surprising, because as far as I know that header isn't even supported on iOS. So please keep me posted and let me know if it indeed solves your issue!

Verstuurd vanaf mijn iPhone

**Comment by markoshust** [comment]
@martijnwalraven thank you, I will. It seems to have solved the issue in my staging environment with TestFlight. I'll confirm with production next week, however it's pretty much the exact same setup, so yes it appears to have resolved it.

**Comment by markoshust** [comment]
@martijnwalraven I managed to execute an update to production (through real App Store), and the hot code push did in fact not execute. The `X-Content-Type-Options: nosniff` command seemed to work with TestFlight, but not through the App Store version. I'm not sure exactly what the difference would be, but I went ahead and reopened this ticket.

**Comment by markoshust** [comment]
I'm getting this error now on production when navigating to `http://mydomain.com/__cordova`

![js-error-3](https://cloud.githubusercontent.com/assets/437029/13555664/c5f76dc0-e394-11e5-82f9-c3829862ad28.png)

**Comment by martijnwalraven** [comment]
The browser errors when navigating to `https://mydomain.com/__cordova` are expected. I just wanted you to check whether it was serving the Cordova specific index page, but it won't actually run correctly in a normal browser (only in a Cordova app, with some additional files and a different downloading mechanism).

Hmmm, so you do have different staging and production servers. Even if they are 1:1 clones, could there be differences in the way the app has been deployed or is configured? Has `ROOT_URL` been set correctly on the production server? Maybe one is using SSL and the other isn't?

**Comment by markoshust** [comment]
Ok thanks for the info.

ROOT_URL has indeed been configured for both. There should be no
differences in either environment, even SSL is on both. Docker + Kubernetes
configs are exactly the same.

I'm going to run TestFlight through the production server and see what
happens.

**Comment by martijnwalraven** [comment]
Are you connecting to both servers through a domain name? Or maybe using an IP address for the staging server? (Just a hunch, this may be nothing.)

**Comment by markoshust** [comment]
Yes, both servers are connected through a domain name.

**Comment by markoshust** [comment]
Hot code push finally worked with my last build/push (beta12)!

I'm guessing there was some sort of problem downloading new assets from the previous build. Is there a guide someplace on how to debug/monitor all of this stuff remotely? I'm a bit in the dark on how to find out what is going on and when.

**Comment by martijnwalraven** [comment]
I'm working on an article on mobile as part of the [Meteor Guide](guide.meteor.com). That should be out around the time 1.3 is released.

**Comment by JacobPariseau** [comment]
Is it likely that the cause of this is just that applications running an older version of Meteor can't receive a hot code push from a new version of Meteor? That seems logical to me, as my hot code pushes broke immediately after updating to 1.3 on mobile only, while locally it worked fine. I'm waiting on Google Play to publish my latest build at which point I expect hot code pushes to begin working again.

**Comment by martijnwalraven** [comment]
@JacobPariseau: Not sure if that was the original issue, but it is true that hot code pushes between 1.2 and 1.3 are intentionally blocked. Because a hot code push only updates JavaScript, and there have been changes to native code between releases, not blocking updates would break the app. This is also true every time you make changes to the set of plugins, so these also require you to submit a new version to the store.
